### ISSUE TITLE
Remove `Storable` in `Foreign.Marshal.Pure` interface

### ISSUE THREAD

**Issue by aspiwack** [tweag/linear-base#11]
I've used `Storable` as a shortcut to implement off-heap data. But this is incorrect on two accounts:

- `Storable` assumes that it has unrestricted argument (in particular for `poke`), whereas we most often have linear values to pass to `poke`.
- `Storable` is in `IO`, which is too unconstrained for our use-case. In principle, `peek` or `poke` could have effects which escape the `unsafePerformIO` call.

I don't have a good solution a replacement type class yet. I'll use this issue to track my thoughts.

**Comment by aspiwack** [comment]
I've looked at the [`storable-record`](https://www.stackage.org/lts-10.4/package/storable-record-0.0.3.1) for inspiration. But it uses non-linearity crucially. I didn't find a way to make it linear yet good.

**Comment by aspiwack** [comment]
I started implementing based on an abstract, authoritative, type class and retractions to the abstract type class. It took much more time than I anticipated to get right. Until I realised there was a significant flaw in my approach. I know a temporary fix. But I will have to move sooner than I excepted to a more generic solution (which I now think should look a bit like the serialised representation example in the paper).

For the moment, I need to take some time away from this.
